test/test_dofs.jl

```diff
@@ -504,3 +504,126 @@ end
+            false    true    true    true 
+        ],
+    ]
+    function is_stored(A, i, j)
+        A = A isa Symmetric ? A.data : A
+        for m in nzrange(A, j)
+            A.rowval[m] == i && return true
+        end
+        return false
+    end
+
+    function check_coupling(dh, topology, K, coupling, elements_coupling)
+        K_check = falses(dh.ndofs[],dh.ndofs[])
+        field_dims = map(fieldname -> Ferrite.getfielddim(dh, fieldname), dh.field_names)
+        for cell_idx in eachindex(getcells(dh.grid))
+            current_face_neighborhood = Ferrite.getdim(dh.grid.cells[cell_idx]) >1 ? topology.face_neighbor[cell_idx,:] : topology.vertex_neighbor[cell_idx,:]
+            current = Ferrite.EntityNeighborhood{FaceIndex}(FaceIndex[FaceIndex((cell_idx, -1))])
+            for neighbor_cell in ∪(current_face_neighborhood,(current,))
+                isempty(neighbor_cell) && continue
+                neighbor_idx = neighbor_cell[1].idx[1]
+                used_coupling = cell_idx == neighbor_idx ? coupling : elements_coupling
+                sz = size(used_coupling, 1)
+                for (fhi, fh) in pairs(dh.fieldhandlers)
+                    for (cell_field_idx, cell_field) in pairs(fh.fields) 
+                        is_discontinuous = Ferrite.IsDiscontinuous(cell_field.interpolation)
+                        cell_field ∈ dh.fieldhandlers[dh.cell_to_fieldhandler[cell_idx]].fields || continue
+                        for (neighbor_field_idx, neighbor_field) in pairs(fh.fields) 
+                            neighbor_field ∈ dh.fieldhandlers[dh.cell_to_fieldhandler[neighbor_idx]].fields || continue
+                            is_discontinuous &= Ferrite.IsDiscontinuous(neighbor_field.interpolation)
+                            for (i_idx, i) in pairs(dof_range(fh, cell_field_idx)), (j_idx, j) in pairs(dof_range(fh,neighbor_field_idx))
+                                I = dh.cell_dofs[i+dh.cell_dofs_offset[cell_idx]-1]
+                                J = dh.cell_dofs[j+dh.cell_dofs_offset[neighbor_idx]-1]
+                                cell_coupling_idx = sz == length(dh.field_names) ? cell_field_idx : (1 + (i_idx - 1) % field_dims[cell_field_idx] + sum(field_dims[1:cell_field_idx-1]))
+                                neighbor_coupling_idx = sz == length(dh.field_names) ? neighbor_field_idx : (1 + (j_idx - 1) % field_dims[neighbor_field_idx] + sum(field_dims[1:neighbor_field_idx-1]))
+                                K_check[I,J] |= (used_coupling[cell_coupling_idx, neighbor_coupling_idx] && (is_discontinuous || (cell_idx == neighbor_idx))) || I == J
+                            end
+                        end
+                    end
+                end
+            end
+        end
+        check_I = repeat(1:dh.ndofs[], dh.ndofs[])
+        check_J = repeat(1:dh.ndofs[], inner = dh.ndofs[])
+        @test is_stored.(Ref(K), check_I, check_J) == reshape(K_check,length(check_I))
+    end
```

Comment:
I'm not sure but the test looks like it's just repeating the implementation in a less optimized way. I'm thinking of a way to automate it.

Reply:
Indeed. Unit tests should never mirror internals, but check for simple invariants.